File: output/GPT4o-Mini/xCodeEval_remove_4000/Python/Go/5cb6c1b3fc0e089985a249812459323b.go
package main

import (
	"bytes"
	"fmt"
	"io"
	"os"
	"sort"
)

const mod = 1e9 + 7
const mod1 = 998244353

type FastIO struct {
	buf    bytes.Buffer
	writer io.Writer
}

func NewFastIO(w io.Writer) *FastIO {
	return &FastIO{writer: w}
}

func (f *FastIO) ReadString() string {
	var s string
	fmt.Fscan(f.writer, &s)
	return s
}

func (f *FastIO) ReadInt() int {
	var n int
	fmt.Fscan(f.writer, &n)
	return n
}

func (f *FastIO) WriteString(s string) {
	f.buf.WriteString(s)
}

func (f *FastIO) WriteInt(n int) {
	f.buf.WriteString(fmt.Sprintf("%d", n))
}

func (f *FastIO) Flush() {
	f.writer.Write(f.buf.Bytes())
	f.buf.Reset()
}

func id9(arr []int, n int, key int) int {
	left, right := 0, n-1
	count := 0

	for left <= right {
		mid := (right + left) / 2
		if arr[mid] <= key {
			count = mid + 1
			left = mid + 1
		} else {
			right = mid - 1
		}
	}
	return count
}

func main() {
	fastIO := NewFastIO(os.Stdout)
	defer fastIO.Flush()

	n := fastIO.ReadInt()
	k := fastIO.ReadInt()
	l := make([]int, n)
	for i := 0; i < n; i++ {
		l[i] = fastIO.ReadInt() % k
	}

	if n == 1 {
		fastIO.WriteInt(l[0])
		return
	}

	l1 := l[:n/2]
	l2 := l[n/2:]

	s := make(map[int]struct{})
	s1 := make(map[int]struct{})
	s2 := make(map[int]struct{})

	for _, v := range l1 {
		for j := range s {
			s2[(j+v)%k] = struct{}{}
		}
		s[v] = struct{}{}
		for j := range s2 {
			s[j] = struct{}{}
		}
		s2 = make(map[int]struct{})
	}

	for _, v := range l2 {
		for j := range s1 {
			s2[(j+v)%k] = struct{}{}
		}
		s1[v] = struct{}{}
		for j := range s2 {
			s1[j] = struct{}{}
		}
		s2 = make(map[int]struct{})
	}

	sList := make([]int, 0, len(s))
	for k := range s {
		sList = append(sList, k)
	}
	sort.Ints(sList)

	s1List := make([]int, 0, len(s1))
	for k := range s1 {
		s1List = append(s1List, k)
	}
	sort.Ints(s1List)

	ans := max(max(sList...), max(s1List...))

	for _, v := range sList {
		t := k - 1 - v
		ind := id9(s1List, len(s1List), t)
		r := s1List[ind-1]
		rt := s1List[len(s1List)-1]
		t = ans
		ans = max((r+v)%k, (v+rt)%k)
		ans = max(ans, t)
	}

	fastIO.WriteInt(ans)
}

func max(a ...int) int {
	m := a[0]
	for _, v := range a {
		if v > m {
			m = v
		}
	}
	return m
}

File: output/GPT3.5-Turbo/xCodeEval_4000/Python/C/5cb6c1b3fc0e089985a249812459323b.c
#include <stdio.h>
#include <stdlib.h>

int binarySearchCount(int arr[], int n, int key) {
    int left = 0;
    int right = n - 1;
    int count = 0;

    while (left <= right) {
        int mid = (right + left) / 2;

        if (arr[mid] <= key) {
            count = mid + 1;
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }

    return count;
}

int main() {
    int n, k;
    scanf("%d %d", &n, &k);
    int *l = (int *)malloc(n * sizeof(int));
    for (int i = 0; i < n; i++) {
        scanf("%d", &l[i]);
        l[i] %= k;
    }

    if (n == 1) {
        printf("%d\n", l[0]);
        return 0;
    }

    int *l1 = (int *)malloc((n / 2) * sizeof(int));
    int *l2 = (int *)malloc((n - n / 2) * sizeof(int));
    for (int i = 0; i < n / 2; i++) {
        l1[i] = l[i];
    }
    for (int i = 0; i < n - n / 2; i++) {
        l2[i] = l[n / 2 + i];
    }

    int *s = (int *)malloc((1 << (n / 2)) * sizeof(int));
    int *s1 = (int *)malloc((1 << (n - n / 2)) * sizeof(int));
    int *s2 = (int *)malloc((1 << (n - n / 2)) * sizeof(int));
    int s_size = 0;
    int s1_size = 0;
    int s2_size = 0;

    for (int i = 0; i < n / 2; i++) {
        for (int j = 0; j < s_size; j++) {
            s2[s2_size++] = (s[j] + l1[i]) % k;
        }
        s[s_size++] = l1[i];
        for (int j = 0; j < s2_size; j++) {
            s[s_size++] = s2[j];
        }
        s2_size = 0;
    }

    for (int i = 0; i < n - n / 2; i++) {
        for (int j = 0; j < s1_size; j++) {
            s2[s2_size++] = (s1[j] + l2[i]) % k;
        }
        s1[s1_size++] = l2[i];
        for (int j = 0; j < s2_size; j++) {
            s1[s1_size++] = s2[j];
        }
        s2_size = 0;
    }

    for (int i = 0; i < s_size; i++) {
        for (int j = 0; j < s1_size; j++) {
            int t = k - 1 - s[i];
            int ind = binarySearchCount(s1, s1_size, t);
            int r = s1[ind - 1];
            int rt = s1[s1_size - 1];
            int t = ans;
            ans = max((r + s[i]) % k, (s[i] + rt) % k);
            ans = max(ans, t);
        }
    }

    printf("%d\n", ans);

    free(l);
    free(l1);
    free(l2);
    free(s);
    free(s1);
    free(s2);

    return 0;
}
